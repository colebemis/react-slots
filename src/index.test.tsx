import { expect, it } from "@jest/globals";
import { render } from "@testing-library/react";
import * as React from "react";
import { useSlots } from ".";

function Layout({ children }: { children?: React.ReactNode }) {
  const [slots, rest] = useSlots(children, {
    header: Header,
    footer: Footer,
  });

  return (
    <div>
      {slots.header
        ? // Pass props to slot with cloneElement()
          React.cloneElement(slots.header, { padding: "large" })
        : null}
      <main>{rest}</main>
      {slots.footer}
    </div>
  );
}

const paddingMap = {
  none: 0,
  small: 4,
  medium: 8,
  large: 16,
};

function Header({ padding = "none" }: { padding?: keyof typeof paddingMap }) {
  return <header style={{ padding: paddingMap[padding] }}>Header</header>;
}

function Footer({ padding = "none" }: { padding?: keyof typeof paddingMap }) {
  return <footer style={{ padding: paddingMap[padding] }}>Footer</footer>;
}

function App() {
  return (
    <Layout>
      <Footer padding="small" />
      <h1>Hello, slots!</h1>
      <Header />
    </Layout>
  );
}

it("works as expected", () => {
  const { container } = render(<App />);
  expect(container).toMatchSnapshot();
});
