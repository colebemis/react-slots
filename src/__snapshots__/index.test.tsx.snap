// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works as expected 1`] = `
<div>
  <div>
    <header
      style="padding: 16px;"
    >
      Header
    </header>
    <main>
      <h1>
        Hello, slots!
      </h1>
    </main>
    <footer
      style="padding: 4px;"
    >
      Footer
    </footer>
  </div>
</div>
`;
